worker/handler/handler.go

```diff
@@ -106,24 +115,92 @@ func NewHandlerManagerSet(opts *config.Config) (hms *HandlerManagerSet, err erro
 	var ihits int64 = 0
 	var misses int64 = 0
 	hms = &HandlerManagerSet{
-		hmMap:      make(map[string]*HandlerManager),
-		regMgr:     rm,
-		pipMgr:     pm,
-		sbFactory:  sf,
-		cacheMgr:   cm,
-		config:     opts,
-		workerDir:  opts.Worker_dir,
-		maxRunners: opts.Max_runners,
-		hhits:      &hhits,
-		ihits:      &ihits,
-		misses:     &misses,
+		hmMap:              make(map[string]*HandlerManager),
+		regMgr:             rm,
+		pipMgr:             pm,
+		sbFactory:          sf,
+		cacheMgr:           cm,
+		config:             opts,
+		workerDir:          opts.Worker_dir,
+		maxRunners:         opts.Max_runners,
+		hhits:              &hhits,
+		ihits:              &ihits,
+		misses:             &misses,
+		sbHardLimit:        opts.Sandbox_create_hard_limit * 1024,
+		numSbCreating:      0,
 	}
 
+	hms.sbCreatingQueue = make(chan *Handler, opts.Sandbox_create_queue_size)
+
 	hms.lru = NewHandlerLRU(hms, opts.Handler_cache_size) //kb
 
+	go hms.updateAvgHandlerMemUsage(opts.Handler_usage_update_interval)
+
+	go hms.scheduleSbCreatingReqs()
+
 	return hms, nil
 }
 
+var memUsage int = 0
+var memLastUpdateTime time.Time = time.Now()
+
+func getMemUsage() int {
+	if (time.Since(memLastUpdateTime)) > time.Second {
+		v, _ := mem.VirtualMemory()
+		memUsage = int(v.Used)
+		memLastUpdateTime = time.Now()
+	}
+	return memUsage
+}
+
+func (hms *HandlerManagerSet) updateAvgHandlerMemUsage(updateInterval int) {
+	for {
+		numHandler := 0
+		totalMemUsage := 0
+		hms.mutex.Lock()
+		startTime := time.Now()
+		for _, hm := range hms.hmMap {
+			for e := hm.handlers.Front(); e != nil; e = e.Next() {
+				h := e.Value.(*Handler)
+				if h.sandbox == nil {
+					continue
+				} else if sbState, err := h.sandbox.State(); err != nil {
+					continue
+				} else if sbState == state.Stopped {
+					continue
+				}
+				numHandler++
+				totalMemUsage += handlerUsage(h)
+			}
+		}
+		usedTime := time.Now().Sub(startTime)
+		if (usedTime > 10 * time.Millisecond) {
+			log.Printf("WARNING: Updating average handler memory usage took more than 10ms (actual time: %v)", usedTime)
+		}
+		hms.mutex.Unlock()
+		if numHandler != 0 {
+			hms.avgHandlerMemUsage = totalMemUsage / numHandler
+		}
+		time.Sleep(time.Duration(updateInterval) * time.Second)
+	}
+}
+
+func (hms *HandlerManagerSet) scheduleSbCreatingReqs() {
```

Comment:
I think this is important enough to warrant some commenting

Reply:
Added comments in the new commit.